controller.go

```diff
@@ -170,60 +172,78 @@ func (n *NodeController) nodeReady(node *v1.Node) (bool, error) {
 // setNodeReady sets node taint macthing ready value. E.g. sets NotReady taint
 // if ready is false, and removes the taint (if exists) when ready is true.
 func (n *NodeController) setNodeReady(node *v1.Node, ready bool) error {
-	// if ready, remove notReady taint if exists on the node
-	if ready {
-		var newTaints []v1.Taint
-		for _, taint := range node.Spec.Taints {
-			if taint.Key != n.taintNodeNotReadyName {
-				newTaints = append(newTaints, taint)
+	setNodeReadiness := func() error {
+		updatedNode, err := n.CoreV1().Nodes().Get(node.Name, metav1.GetOptions{})
+		if err != nil {
+			return backoff.Permanent(err)
+		}
+
+		// if ready, remove notReady taint if exists on the node
+		if ready {
+			var newTaints []v1.Taint
+			for _, taint := range updatedNode.Spec.Taints {
+				if taint.Key != n.taintNodeNotReadyName {
+					newTaints = append(newTaints, taint)
+				}
+			}
+
+			if len(newTaints) == len(updatedNode.Spec.Taints) {
+				return nil
+			}
+			updatedNode.Spec.Taints = newTaints
+		} else {
+			if hasTaint(updatedNode, n.taintNodeNotReadyName) {
+				return nil
+			}
+
+			taint := v1.Taint{
+				Key:    n.taintNodeNotReadyName,
+				Effect: v1.TaintEffectNoSchedule,
 			}
+			updatedNode.Spec.Taints = append(updatedNode.Spec.Taints, taint)
 		}
 
-		if len(newTaints) != len(node.Spec.Taints) {
-			node.Spec.Taints = newTaints
-			_, err := n.CoreV1().Nodes().Update(node)
-			if err != nil {
+		_, err = n.CoreV1().Nodes().Update(updatedNode)
+		if err != nil {
+			// automatically retry if there was a conflicting update.
+			if errors.IsConflict(err) {
 				return err
 			}
+			return backoff.Permanent(err)
+		}
+
+		if ready {
 			log.WithFields(log.Fields{
 				"action": "removed",
 				"taint":  n.taintNodeNotReadyName,
-				"node":   node.ObjectMeta.Name,
+				"node":   updatedNode.ObjectMeta.Name,
 			}).Info("")
 
 			if n.nodeStartUpObeserver != nil {
```

Comment:
nit: observer

Reply:
This was auto-completed wrong all over the place. Fixed now, thanks!

Reply:
One thing you could do is have the function accept `nodeName string` to avoid autocomplete giving you the wrong object.